### ISSUE TITLE
[go_router] Nested ShellRoute - Unable to exit app

### ISSUE THREAD

**Issue by NeroThroN** [flutter/flutter#128793]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

The application has a nested ShellRoute configuration: 

The first ShellRoute is here to display a loading indication while some async method load. It's a ShellRoute because it has to access NavigatorState inside the load function to launch `context.go('/onboarding')` for example. [Doc](https://docs.google.com/document/d/10l22o4ml4Ss83UyzqUC8_xYOv_QjZEi80lJDNE4q7wM/edit?pli=1&resourcekey=0-U-BXBQzNfkk4v241Ow-vZg)

The second ShellRoute is used to display a bottom navigation bar or any other navigation widget on top of the application's screens. This navigation's ShellRoute is used only for the '/' in the example provided below.

### Expected results

1. Try to exit the application by using Android back button => Exit application 
2. Try to exit the application by using "Go back" button => Exit application

### Actual results

1. Try to exit the application by using Android back button => Nothing seems to happen
2. Try to exit the application by using "Go back" button => **`Exception has occured:`** `LateError (LateInitializationError: Local 'subRoute' has not been initialized.)`

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() => runApp(const MainApp());

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp.router(routerConfig: appRouter);
}

final rootKey = GlobalKey<NavigatorState>(debugLabel: 'loading');
final navKey = GlobalKey<NavigatorState>(debugLabel: 'navigation');

final appRouter = GoRouter(
  debugLogDiagnostics: true,
  routes: [
    ShellRoute(
      navigatorKey: rootKey,
      builder: (context, state, child) => BasicLayout(name: 'Loading', child: child),
      routes: [
        ShellRoute(
          navigatorKey: navKey,
          builder: (context, state, child) => BasicLayout(name: 'Navigation', child: child),
          routes: [
            GoRoute(path: '/', parentNavigatorKey: navKey, builder: (context, state) => const BasicScreen(name: 'Home'), routes: [
              GoRoute(path: 'settings', parentNavigatorKey: rootKey, builder: (context, state) => const BasicScreen(name: 'Settings')),
            ]),
          ],
        ),
        GoRoute(path: '/deeplink', parentNavigatorKey: rootKey, redirect: (_, __) => '/?deeplink=true'),
        GoRoute(path: '/onboarding', parentNavigatorKey: rootKey, builder: (context, state) => const BasicScreen(name: 'Onboarding')),
      ],
    ),
  ],
);

class BasicLayout extends StatelessWidget {
  const BasicLayout({super.key, required this.child, required this.name});

  final Widget child;
  final String name;

  @override
  Widget build(BuildContext context) => Scaffold(appBar: AppBar(title: Text('Layout: $name')), body: child);
}

class BasicScreen extends StatelessWidget {
  const BasicScreen({super.key, required this.name});

  final String name;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Screen: $name')),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ElevatedButton(
              onPressed: context.pop,
              child: const Text('Go back'),
            ),
            ElevatedButton(
              onPressed: () => context.go('/settings'),
              child: const Text('Settings'),
            )
          ],
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.4, on Fedora Linux 38 (Workstation Edition) 6.3.6-200.fc38.x86_64, locale en_US.UTF-8)
    • Flutter version 3.10.4 on channel stable at /home/nerothron/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 682aa387cf (8 days ago), 2023-06-05 18:04:56 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/nerothron/Android/Sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /var/lib/snapd/snap/android-studio/126/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /var/lib/flatpak/exports/bin/com.google.Chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2022.2)
    • Android Studio at /var/lib/snapd/snap/android-studio/126/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.79.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.66.0

[✓] Connected device (3 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 12 (API 31) (emulator)
    • Linux (desktop)              • linux         • linux-x64      • Fedora Linux 38 (Workstation Edition) 6.3.6-200.fc38.x86_64
    • Chrome (web)                 • chrome        • web-javascript • Google Chrome 113.0.5672.126 unknown

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by dam-ease** [comment]
Hi @NeroThroN. Thanks for filing this issue. I am able to reproduce this on the latest `master` and `stable` channels. On the latest `stable` channel, android back button exits the application but the exception with pressing the Go back button still occurs.

<details><summary>flutter doctor -v</summary>
<p>

```
[✓] Flutter (Channel stable, 3.10.4, on macOS 13.4 22F66 darwin-arm64, locale
    en-NG)
    • Flutter version 3.10.4 on channel stable at /Users/dammya/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 682aa387cf (8 days ago), 2023-06-05 18:04:56 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version
    32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.78.0)
    • VS Code at /Users/dammya/.Trash/Visual Studio Code 20.07.05.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (4 available)
    • sdk gphone arm64 (mobile) • emulator-5554                        •
      android-arm64  • Android 11 (API 30) (emulator)
    • iPhone 14 (mobile)        • 5FE8DE93-0F4C-45BA-8BED-BC9CD69B35D1 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)           • macos                                •
      darwin-arm64   • macOS 13.4 22F66 darwin-arm64
    • Chrome (web)              • chrome                               •
      web-javascript • Google Chrome 114.0.5735.106

[✓] Network resources
    • All expected network resources are available.

• No issues found!
``` 

```
[✓] Flutter (Channel master, 3.12.0-4.0.pre.60, on macOS 13.4 22F66 darwin-arm64, locale en-NG)
    • Flutter version 3.12.0-4.0.pre.60 on channel master at /Users/dammya/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4e706975b7 (35 minutes ago), 2023-06-13 08:42:21 -0400
    • Engine revision f9a0a0dafe
    • Dart version 3.1.0 (build 3.1.0-204.0.dev)
    • DevTools version 2.24.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    ✗ CocoaPods installed but not working.
        You appear to have CocoaPods installed but it is not working.
        This can happen if the version of Ruby that CocoaPods was installed with is different from the one being used to invoke it.
        This can usually be fixed by re-installing CocoaPods.
      To re-install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.78.0)
    • VS Code at /Users/dammya/.Trash/Visual Studio Code 20.07.05.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 13.4 22F66 darwin-arm64
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 114.0.5735.106

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

``` 
</p>
</details>

**Comment by leanhro2812** [comment]
Tks, Fixed on go_router 8.1.0
